lib/es6-promise/promise/resolve.js

```diff
@@ -38,7 +38,7 @@ export default function resolve(object) {
   /*jshint validthis:true */
   let Constructor = this;
 
-  if (object && typeof object === 'object' && object.constructor === Constructor) {
+  if (object instanceof Constructor) {
```

Comment:
To support spec compliant subclassing, this must be exact constructor match.

(If tests pass with this, we likely need to expand our test coverage)